src/dev/run_prepush_hook.ts

```diff
@@ -0,0 +1,278 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { existsSync } from 'node:fs';
+import { simpleGit } from 'simple-git';
+import cliProgress from 'cli-progress';
+import chalk from 'chalk';
+import Table from 'cli-table3';
+import { run } from '@kbn/dev-cli-runner';
+import { REPO_ROOT } from '@kbn/repo-info';
+import { ToolingLog } from '@kbn/tooling-log';
+import { getTimeReporter } from '@kbn/ci-stats-reporter';
+
+import { File } from './file';
+import { checkFileCasing } from './preflight_check/check_file_casing';
+import { checkTypescriptFiles } from './preflight_check/run_tsc';
+import { runUnitTests } from './preflight_check/run_unit_tests';
+import { esLintFiles } from './preflight_check/run_eslint';
+
+const HASH_FOR_NULL_OR_DELETED_FILE = '00000000000';
+
+function getDefaults(taskName: string) {
+  const TASK_COL_LENGTH = 10;
+  return {
+    task: chalk.blue(taskName.padEnd(TASK_COL_LENGTH)),
+    filename: '',
+  };
+}
+
+function nonNullable<T>(value: T): value is NonNullable<T> {
+  return value != null;
+}
+
+run(
+  async ({ log, flags }) => {
+    const toolingLog = new ToolingLog({
+      level: 'info',
+      writeTo: process.stdout,
+    });
+
+    const runStartTime = Date.now();
+    const reportTime = getTimeReporter(toolingLog, 'scripts/run_prepush_hook');
+
+    process.env.IS_KIBANA_PREPUSH_HOOK = 'true';
+
+    const git = simpleGit(REPO_ROOT);
+
+    if (!(await git.branchLocal().status()).isClean()) {
+      reportTime(runStartTime, 'total', {
+        success: false,
+      });
+
+      throw new Error(
+        'Local files have changes that are not committed yet. Skipping preflight checks.'
+      );
+    }
+
+    const checkTypes = ['i18n', 'tsc', 'eslint', 'jest', 'fileCasing'] as const;
+
+    const checks = checkTypes.reduce((acc, check) => {
+      acc[check] = {
+        files: [],
+      };
+
+      return acc;
+    }, {} as Record<typeof checkTypes[number], { files: Array<{ path: string; file: File }> }>);
+
+    const { current } = await git.branchLocal();
+
+    const commonAncestor = (await git.raw(['merge-base', 'origin/main', current])).trim();
```

Comment:
I'm not sure why, but when I check out this PR locally and run `node scripts/prepush_hook -v`, I would expect it to only check the files edited in this PR. However, it seems to check thousands of files (I've never been patient enough to wait for it to finish, and so I have to `kill -9` it).

Reply:
Ah!! I found the issue: The problem is that I call the remote for `upstream` and not `origin`. I do have one called `origin`, but it's several months old, hence the enormous change-set.

Reply:
Good find! Need a way to determine the upstream's name on the local install in a consistent way. Maybe it's possible to just use the Kibana repo link directly, but didn't succeed with that yet.

Reply:
This is my recommended fix, which also reduces the number of commands to run:

```diff
-const { current } = await git.branchLocal();
-
-const commonAncestor = (await git.raw(['merge-base', 'origin/main', current])).trim();
-
-const diff = await git.diff([`${commonAncestor}..${current}`]);
+const diff = await git.diff(['--merge-base', 'main']);
```

This changes 3 things:

1. No need to get the current branch. You can either just use `HEAD` or leave it out as it's the default
2. No need to manually get the merge-base. The diff command supports this using the `--merge-base` argument
3. No need to know the name of the remote. You can just leave it out and rely on the local `main` branch.

Reply:
One remaining issue here is that we technically don't know if the target branch is going to be `main`. It could be that the target branch is a feature branch. So it would be nice if it was somehow possible to specify the target branch somehow (e.g. a command line argument or something like that).

Reply:
Using

```const diff = await git.diff(['--merge-base', 'main']);```

makes this the script report this PR as 97 (and after merging in upstream 1024) files, whereas it should be 14.

Reply:
Interesting problem. I assume this is because you've merged in `main` to the PR but haven't yet pull down `main` to your local clone. So it has to go further back to find the common ancestor. But I also assume that the reverse could happen as well if you haven't fetched from `origin` in a while (as I don't assume `git diff --merge-base origin/main` actually makes any network requests, but just look in `.git/refs/remotes/origin/main`?)

Reply:
I guess one algorithm that might "solve" this dilemma is running two diffs: One against `main` and another against `origin/main` and then pick the one with the smallest number of commits in-between.

This however, re-introduces the issue of knowing the name of the remote to use

Reply:
> (as I don't assume git diff --merge-base origin/main actually makes any network requests, but just look in .git/refs/remotes/origin/main?)

I just checked by running `git show upstream/main` locally and as I expected it's using `.git/refs/remotes/origin/main` - so if you've merged in `main` to your PR on GitHub and then pulled down the changes to the PR branch locally, I would expect you would have the same issue.

Reply:
I see what you mean. Not fully sure what the best way forward is here yet.